contracts/libs/MarginAccount.sol

```diff
@@ -1,12 +1,220 @@
+     * @param _index The index of _shortOtoken in the user's vault.shortOtokens array
+     */
+    function _removeShort(
+        Vault storage _vault,
+        address _shortOtoken,
+        uint256 _amount,
+        uint256 _index
+    ) internal {
+        //Check that the removed short otoken exists in the vault
+        require(
+            (_index < _vault.shortOtokens.length) && (_vault.shortOtokens[_index] == _shortOtoken),
+            "MarginAccount: short otoken address mismatch"
+        );
+
+        _vault.shortAmounts[_index] = _vault.shortAmounts[_index].sub(_amount);
+    }
+
+    /**
+     * @dev increase the long oToken balance in a vault when an oToken is deposited
+     * @param _vault The vault the protocol is adding the long position to
+     * @param _longOtoken The address of the _longOtoken the protocol is adding to the user's vault
+     * @param _amount The number of _longOtoken the protocol is adding to the user's vault
+     * @param _index The index of _longOtoken in the user's vault.longOtokens array
+     */
+    function _addLong(
+        Vault storage _vault,
+        address _longOtoken,
+        uint256 _amount,
+        uint256 _index
+    ) internal {
+        // Valid indexes in any array are between 0 and array.length - 1.
+        // If adding amount into an existant long otoken, make sure _index in the range of 0->length-1
+        if ((_index >= _vault.longOtokens.length) && ((_index >= _vault.longAmounts.length))) {
+            _vault.longOtokens.push(_longOtoken);
+            _vault.longAmounts.push(_amount);
+        } else {
+            require(
+                (_vault.longOtokens[_index] == _longOtoken) || (_vault.longOtokens[_index] == address(0)),
+                "MarginAccount: invalid long otoken position"
+            );
+
+            _vault.longAmounts[_index] = _vault.longAmounts[_index].add(_amount);
+            _vault.longOtokens[_index] = _longOtoken;
+        }
+    }
+
+    /**
+     * @dev decrease the long oToken balance in a vault when an oToken is withdrawn
+     * @param _vault The vault from which the protocol is decreasing the short position
+     * @param _longOtoken The address of the _longOtoken of which the protocol is reducing the long position from the user's vault
+     * @param _amount The number of _longOtoken the protocol is reducing the user's long position by from the user's vault
+     * @param _index The index of _longOtoken in the user's vault.longOtokens array
+     */
+    function _removeLong(
+        Vault storage _vault,
+        address _longOtoken,
+        uint256 _amount,
+        uint256 _index
+    ) internal {
+        //Check that the removed long token exists in the vault
+        require(
+            (_index < _vault.longOtokens.length) && (_vault.longOtokens[_index] == _longOtoken),
+            "MarginAccount: long otoken address mismatch"
+        );
+
+        _vault.longAmounts[_index] = _vault.longAmounts[_index].sub(_amount);
```

Comment:
if after removing long, `_vault. longAmounts[_index] == 0`, should we remove the long otoken address from `_vault.longOtokens[_index]` ?! @zkoticha @aparnakr @antoncoding

Reply:
@aparnakr @antoncoding

Reply:
sure, that seems good